### DISCUSSION TITLE
[@types/react]  does not exist on type 'JSX.IntrinsicElements'.ts  after upgrade to react 19 and  @types/react 19.0.1

### DISCUSSION THREAD

**Discussion by lichuhub** [DefinitelyTyped/DefinitelyTyped#71395] [Issues with a @types package] [answered]
No issue after install react 19.  But once we upgraded to @types/react 19.0.1 we are getting   does not exist on type 'JSX.IntrinsicElements'.ts error. Downgrade to "18.3.14" the error goes away.

Code example.
declare global {
    namespace JSX {
        interface IntrinsicElements {
            "shell_layout": DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>;
            "shell_breadcrumb": DetailedHTMLProps<
                HTMLAttributes<HTMLElement>,
                HTMLElement
            >;
        }
    }
}

const Layout = () => {
    return (
        <shell_layout scope-query-selector="shell_layout" slot="header">
            <shell_breadcrumb id="extra-crumbs" />
            <LeftNavigation  />
            <Outlet />
        </shell_layout>
    );
};

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment, marked as answer]
The JSX namespace is no longer global but scoped to React. You'll now need to wrap it in `declare module "...."` instead of `declare global`:

```diff
// global.d.ts

- declare global {
+ declare module "react" {
    namespace JSX {
      interface IntrinsicElements {
        "my-element": {
          myElementProps: string;
        };
      }
    }
+ }
```

The exact module specifier depends on the JSX runtime you specified in the `compilerOptions` of your `tsconfig.json`.
For `"jsx": "react-jsx"` it would be `react/jsx-runtime`.
For `"jsx": "react-jsxdev"` it would be `react/jsx-dev-runtime`.
For `"jsx": "react"` and `"jsx": "preserve"` it would be `react`.

**Reply by sagrawal31** [reply]
For me, I had `"jsx": "react-jsx",` so writing the following worked for me-

```
declare module 'preact' {
    namespace JSX {
        // https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/71395
        interface IntrinsicElements {
            'cooee-product-options-picker': {
                item: Item;
                element: DropdownElement,
            };
        }
    }
}
```

**Comment by lichuhub** [comment]
I see, i will try that later today. 
Edit: after adding that. Now i got Property 'p' 'div' does not exist on type 'JSX.IntrinsicElements'.ts on other pages.

I was using "jsx": "preserve".

  ok I solved the issue by not adding the type to tsconfig.json instead. I just added to the component that's going to use it.

**Reply by mrdjohnson** [reply]
> Now i got Property 'p' 'div' does not exist on type 'JSX.IntrinsicElements'.ts on other pages.

I got a similar result. I had to import react on file for some reason

// global.d.ts or whatever your file name is, mine is src/types/JSX.ts
```
import 'react'

declare module 'react/jsx-runtime' {
  namespace JSX {
    interface IntrinsicElements {
      math: any
    }
  }
}
```

follow up: while the above fixed ONE of my `IntrinsicElements` TS errors (`<math />` for example) react-markdown was still mad at me; here is what I did that actually fixed everything 

```
declare namespace React {
  namespace JSX {
    interface IntrinsicElements {
      math: any
    }
  }
}
```

**Reply by TheWalkingDad3** [reply]
thx a lot, you saved me!

When I used the code from the react docu:
```
declare module "react" {
    namespace JSX {
      interface IntrinsicElements {
        "my-element": {
          myElementProps: string;
        };
      }
    }
}
```

all react imports in other files did not work any longer.

But when I added `import 'react'` to my typing definitions file everything works great :)

**Comment by marcosfilho-hotmart** [comment]
I had a similar issue when declaring the interface on a "clean" file, or any file that has no other type or interface declaration, but the interface declaration worked just fine when it was near other type declaractions:

The following code erros:
```typescript
// global.d.ts (start of file)

declare module 'react/jsx-runtime' {
  namespace JSX {
    interface IntrinsicElements {
      'my-tag-name': TagType
      'my-other-tag': OtherTagType
      ...
    }
  }
}
// End of file
```

The following code works without errors
```typescript
// typings.ts

export interface BundleProps {
  language: AvailableLanguages
  user: User
}

declare module 'react/jsx-runtime' {
  namespace JSX {
    interface IntrinsicElements {
      'my-tag-name': TagType
      'my-other-tag': OtherTagType
      ...
    }
  }
}
```

**Comment by eliotgevers** [comment]
I got it working with Next.js 16 and React 19.2. Here are my versions:

"next": "16.0.1", "react": "19.2.0", "react-dom": "19.2.0", "three": "^0.181.0", "@types/three": "^0.181.0", "@react-three/fiber": "^9.4.0"

With the following globals.d.ts setup:

```
import { ThreeElements } from "@react-three/fiber";

declare module "react/jsx-runtime" {
  namespace JSX {
    interface IntrinsicElements extends ThreeElements {}
  }
}
```

It still shows the following linting error, but everything works: "An interface declaring no members is equivalent to its supertype."

Restart your TS server while trying this, because it sometimes shows outdated information.
Make sure to install the @types/three package
Also that the globals.d.ts file is found by tsconfig (either in src folder, or add it to the "include")
